Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,202 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` an `(R, S)`-bimodule.
```

Comment:
Your assumptions about the definition of "bimodule" in this file are not consistent with e.g. the use of the term in `Algebra.Module.Bimodule`.

Reply:
In the latest change, I think I have updated this file so that the bimodule is actually mathematically correct.

Reply:
I think this has just substantially confused the issue; If we're using `Sᵐᵒᵖ` everywhere and `S` nowhere, then we've sacrificed generality and forced ourselves onto a more fragile end of mathlib (c.f. #7152).

Reply:
If I understand this issue correctly, then we have three options

1. assume `X` is both a (left) `R`-module and a (left) `S`-module such that `R`-action and `S`-action commutes. Then define everything in terms of left modules. This would not technically be the tensor-hom adjunction, but it is easier to work with and to maintain (at least for now, maybe situation would improve after #7152)
2. assume `X` is an `(R, S)`-bimodule and define the tensor-hom adjunction as an adjunction between left `R`-module and right `S`-module. This would force us to use `Module S^{mop} ...` and mixing left module and right module is awkward.
3. assume `R, S` are both commutative and pretend that left and right module makes no difference